modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,320 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+from modin.error_message import ErrorMessage
+from modin.core.dataframe.base.dataframe.utils import Axis
+from typing import Union, Callable, Optional
+from modin.config import NPartitions
+import ray
+import numpy as np
+from modin.distributed.dataframe.pandas.partitions import from_partitions
+
+
+class PandasQuery(object):
+    def __init__(
+        self,
+        func: Callable,
+        is_output: bool = False,
+        repartition_after: bool = False,
+        fan_out: bool = False,
+        pass_partition_id: bool = False,
+        reduce_fn: Optional[Callable] = None,
+    ):
+        self.func = func
+        self.is_output = is_output
+        self.oids = []
+        self.fan_out = fan_out
+        self.reptn = repartition_after
+        self.pass_num_ptn = pass_partition_id
+        self.operators = None
+        self.reduce_fn = reduce_fn
+
+    @classmethod
+    def print_options(cls):
+        """
+        Prints the kwargs that Query objects accept.
+        """
+        print(
+            """is_output : bool, default: False
+    Whether this query is an output query and should be passed both to the next query, and directly to postprocessing.
+repartition_after : bool, default: False
+    Whether to repartition after this query is computed. Currently, repartitioning is only supported if there is 1 partition prior to repartitioning.
+fan_out : bool, default: False
+    Whether to fan out this node. If True and only 1 partition is passed as input, the partition is replicated `num_partition` times, and the function is called on each. The `reduce_fn` must also be specified.
+pass_partition_id : bool, default: False
+    Whether to pass the numerical partition id to the query.
+reduce_fn : Callable
+    The reduce function to apply if `fan_out` is set to True. This takes the `num_partition` partitions that result from this query, and combines them into 1 partition.
+output_id : int, default: None
+    An id to assign to this node if it is an output."""
+        )
+
+
+class PandasQueryPipeline(object):
```

Comment:
We need some defensive coding here:
* Check on the `Engine.get() == "ray"`
* Check on `df` that it is backed by ray